File: app/src/main/java/nl/jpelgrm/movienotifier/data/WatcherListDiffCallback.java
package nl.jpelgrm.movienotifier.data;

import android.support.v7.util.DiffUtil;

import java.util.List;

import nl.jpelgrm.movienotifier.models.Watcher;

public class WatcherListDiffCallback extends DiffUtil.Callback {
    private List<Watcher> oldList;
    private List<Watcher> newList;

    public WatcherListDiffCallback(List<Watcher> oldList, List<Watcher> newList) {
        this.oldList = oldList;
        this.newList = newList;
    }

    @Override
    public int getOldListSize() {
        return oldList.size();
    }

    @Override
    public int getNewListSize() {
        return newList.size();
    }

    @Override
    public boolean areItemsTheSame(int oldItemPosition, int newItemPosition) {
        return oldList.get(oldItemPosition).getID().equals(newList.get(newItemPosition).getID());
    }

    @Override
    public boolean areContentsTheSame(int oldItemPosition, int newItemPosition) {
        Watcher oldWatcher = oldList.get(oldItemPosition);
        Watcher newWatcher = newList.get(newItemPosition);

        return oldWatcher != null && newWatcher != null && oldWatcher.equals(newWatcher);
    }
}

File: app/src/main/java/nl/jpelgrm/movienotifier/data/WatcherListDiffCallback.kt
package nl.jpelgrm.movienotifier.data

import androidx.recyclerview.widget.DiffUtil
import nl.jpelgrm.movienotifier.models.Watcher

class WatcherListDiffCallback(private val oldList: List<Watcher>, private val newList: List<Watcher>) : DiffUtil.Callback() {
    override fun getOldListSize(): Int {
        return oldList.size
    }

    override fun getNewListSize(): Int {
        return newList.size
    }

    override fun areItemsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        return oldList[oldItemPosition].id == newList[newItemPosition].id
    }

    override fun areContentsTheSame(oldItemPosition: Int, newItemPosition: Int): Boolean {
        val oldWatcher = oldList[oldItemPosition]
        val newWatcher = newList[newItemPosition]
        return oldWatcher == newWatcher
    }
}